File: python/mirage/commands/builtin/generic_bind/builders/cp.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from dataclasses import replace
from functools import partial

from mirage.accessor.base import Accessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.cp import cp as generic_cp
from mirage.commands.builtin.generic.cp import parse_flags
from mirage.commands.builtin.generic.find import parse_find_args, walk_find
from mirage.commands.builtin.generic_bind.adapter import (Builder, CommandIO,
                                                          Operation, bound_op,
                                                          overlaid_stat)
from mirage.commands.config import CommandOpts
from mirage.commands.spec import SPECS
from mirage.commands.spec.flag_view import FlagView
from mirage.context import hidden_paths_intersect, path_rules_active
from mirage.io.types import ByteSource, IOResult
from mirage.ops.types import StatOverlay
from mirage.types import NativeCopy, PathSpec, PrimitiveCopy
from mirage.utils.key_prefix import rekey
from mirage.vfs.types import OperationFn


async def _walk_find(readdir: OperationFn,
                     stat: OperationFn,
                     index: IndexCacheStore,
                     src: PathSpec,
                     type: str | None = None) -> list[str]:
    results = await walk_find(src,
                              readdir=readdir,
                              stat=stat,
                              index=index,
                              args=parse_find_args((), type=type))
    return ["/" + rekey(src.virtual, src.vfs_path, path) for path in results]


def _make_find(ops: CommandIO, accessor: Accessor,
               index: IndexCacheStore) -> OperationFn:
    if ops.find is not None:
        return partial(ops.find, accessor, index=index)
    return partial(_walk_find, partial(ops.readdir, accessor),
                   partial(ops.stat, accessor), index)


def overlayable_stat(ops: CommandIO, accessor: Accessor,
                     index: IndexCacheStore,
                     stat_overlay: StatOverlay | None) -> OperationFn:
    """The backend stat, merged with the namespace attr overlay if any.

    cp/mv freshness checks (``-u``) must see touch/chmod overlay state,
    exactly like ls and stat rendering.

    Args:
        ops (CommandIO): Backend command IO facade.
        accessor (Accessor): Backend handle.
        index (IndexCacheStore): Cache index threaded through.
        stat_overlay (StatOverlay | None): Namespace merge, or None.
    """
    if stat_overlay is None:
        return bound_op(ops.stat, accessor, index)
    return partial(overlaid_stat,
                   partial(ops.stat, accessor),
                   stat_overlay,
                   index=index)


async def _write(op: OperationFn,
                 accessor: Accessor,
                 path: PathSpec,
                 data: bytes = b"") -> None:
    await op(accessor, path, data)


async def cp(ops: CommandIO, accessor: Accessor, paths: list[PathSpec],
             texts: list[str],
             opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    if not ops.is_mounted(accessor):
        raise ValueError("cp: no VFS")
    fl = FlagView(opts.flags, spec=SPECS["cp"])
    parsed = parse_flags(fl)
    paths = await ops.resolve_glob(accessor, paths, opts.index)
    dir_copy = partial(ops.dir_copy, accessor) if ops.dir_copy else None
    mkdir = partial(ops.mkdir, accessor) if ops.mkdir else None
    if ops.copy is None and ops.write is None:
        # Directory creation is not a usable copy step without a file
        # transfer capability. Refuse it through the same guarded door
        # before the command leaves an uncopyable destination tree.
        mkdir = partial(
            replace(ops, mkdir=None).require(Operation.MKDIR), accessor)
    strategy: NativeCopy | PrimitiveCopy
    guarded = path_rules_active() or any(
        hidden_paths_intersect(p.virtual) for p in paths)
    primitive = ops.copy is None or (guarded and mkdir is not None)
    if primitive and ops.write is not None:
        # A native copy moves a tree in one backend call and a native
        # find lists it, neither of which passes an entry through the
        # guard the way a read does; while a path rule scopes cp, or a
        # hide could cover an entry under an operand (the native find
        # listed hidden names and the per-file read then printed them
        # in its refusal), the primitive walk copies entry by entry
        # (the cross-mount relay's own path), which is also where GNU's
        # per-entry refusals are worded.
        strategy = PrimitiveCopy(read_bytes=bound_op(ops.read_bytes, accessor,
                                                     opts.index),
                                 write=partial(_write, ops.write, accessor),
                                 mkdir=partial(ops.require(Operation.MKDIR),
                                               accessor),
                                 readdir=bound_op(ops.readdir, accessor,
                                                  opts.index))
    else:
        strategy = NativeCopy(copy=partial(ops.require(Operation.COPY),
                                           accessor),
                              find=_make_find(ops, accessor, opts.index),
                              dir_copy=dir_copy,
                              mkdir=mkdir)
    overlay = opts.ns.stat_overlay if opts.ns is not None else None
    return await generic_cp(paths,
                            strategy=strategy,
                            stat=overlayable_stat(ops, accessor, opts.index,
                                                  overlay),
                            flags=parsed,
                            readdir=bound_op(ops.readdir, accessor,
                                             opts.index))


BUILDER = Builder('cp', cp, write=True)

File: typescript/packages/core/src/commands/builtin/generic_bind/builders/cp.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { IndexCacheStore } from '../../../../cache/index/store.ts'
import type { StatOverlay } from '../../../../ops/types.ts'
import type { Accessor } from '../../../../accessor/base.ts'
import type { NativeCopy, PathSpec, PrimitiveCopy, StatFn } from '../../../../types.ts'
import { hiddenPathsIntersect, pathRulesActive } from '../../../../context/session_context.ts'
import { walkFind } from '../../../../core/generic/find.ts'
import { cpGeneric, parseFlags } from '../../generic/cp.ts'
import type { Builder, CommandIO } from '../adapter.ts'
import { requireOp, resolveGlobOf } from '../adapter.ts'
import { FlagView } from '../../../spec/flag_view.ts'
import { specOf } from '../../../spec/builtins.ts'

// The backend stat, merged with the namespace attr overlay if any. cp/mv
// freshness checks (-u) must see touch/chmod overlay state, exactly like
// ls and stat rendering.
export function overlayableStat(
  ops: CommandIO,
  accessor: Accessor,
  index: IndexCacheStore | undefined,
  statOverlay: StatOverlay | undefined,
): StatFn {
  if (statOverlay === undefined) return (p) => ops.stat(accessor, p, index)
  return async (p) => statOverlay(p.virtual, await ops.stat(accessor, p, index))
}

export const CP_BUILDER: Builder = {
  name: 'cp',
  write: true,
  fn: async (ops, accessor, paths, _texts, opts) => {
    const { dirCopy, find } = ops
    // Without a file transfer capability, creating directories would
    // leave an uncopyable destination tree. Keep the refusal guarded.
    const mkdir =
      ops.copy === undefined && ops.write === undefined
        ? requireOp<NonNullable<CommandIO['mkdir']>>(undefined, 'mkdir')
        : ops.mkdir
    const copy = requireOp(ops.copy, 'copy')
    const idx = opts.index ?? undefined
    const resolved = await resolveGlobOf(ops)(accessor, paths, idx)
    // No native find op: fall back to a readdir walk (mirrors Python's
    // _make_find). Passing the index lets stat classify entries from the
    // cache instead of re-fetching, matching the find command.
    const findFn: NativeCopy['find'] =
      find !== undefined
        ? (src, options) => find(accessor, src, options, idx)
        : (src, options) =>
            walkFind(
              src,
              {
                readdir: (spec, i) => ops.readdir(accessor, spec, i),
                stat: (spec, i) => ops.stat(accessor, spec, i),
              },
              options,
              idx,
            )
    const parsed = parseFlags(new FlagView(opts.flags, specOf('cp')))
    // A native copy moves a tree in one backend call and a native find
    // lists it, neither of which passes an entry through the guard the
    // way a read does; while a path rule scopes cp, or a hide could
    // cover an entry under an operand (the native find listed hidden
    // names and the per-file read then printed them in its refusal),
    // the primitive walk copies entry by entry (the cross-mount
    // relay's own path), which is also where GNU's per-entry refusals
    // are worded.
    const { write } = ops
    const guarded = pathRulesActive() || resolved.some((p) => hiddenPathsIntersect(p.virtual))
    const primitive = ops.copy === undefined || (guarded && mkdir !== undefined)
    const strategy: NativeCopy | PrimitiveCopy =
      primitive && write !== undefined
        ? {
            readBytes: (p: PathSpec) => ops.readBytes(accessor, p, idx),
            write: (p: PathSpec, data: Uint8Array) => write(accessor, p, data),
            mkdir: (p: PathSpec) => requireOp(ops.mkdir, 'mkdir')(accessor, p),
            readdir: (p: PathSpec) => ops.readdir(accessor, p, idx),
          }
        : {
            copy: (src: PathSpec, target: PathSpec) => copy(accessor, src, target),
            find: findFn,
            ...(dirCopy === undefined
              ? {}
              : { dirCopy: (src: PathSpec, target: PathSpec) => dirCopy(accessor, src, target) }),
            ...(mkdir === undefined ? {} : { mkdir: (p: PathSpec) => mkdir(accessor, p) }),
          }
    return cpGeneric(
      resolved,
      overlayableStat(ops, accessor, idx, opts.ns?.statOverlay),
      strategy,
      parsed,
      idx,
      undefined,
      (p: PathSpec) => ops.readdir(accessor, p, idx),
    )
  },
}
